lib/auth/auth_dtls.c

```diff
@@ -0,0 +1,959 @@
+		 */
+		LOG_DBG("Waiting for more bytes to fill DTLS packet.");
+	}
+
+	/* Didn't send any bytes */
+	return 0;
+}
+
+
+/**
+ * Set the DTLS cookie.
+ *
+ * @param auth_conn   Pointer to auth connection
+ *
+ * @return 0 on success, else error code.
+ */
+static int auth_tls_set_cookie(struct authenticate_conn *auth_conn)
+{
+	int ret;
+	struct mbed_tls_context *mbed_ctx = (struct mbed_tls_context *)auth_conn->internal_obj;
+
+	/* should not be NULL!!  */
+	if (!mbed_ctx) {
+		LOG_ERR("No MBED context.");
+		return AUTH_ERROR_INVALID_PARAM;
+	}
+
+	ret = mbedtls_ssl_set_client_transport_id(&mbed_ctx->ssl, mbed_ctx->cookie,
+						  sizeof(mbed_ctx->cookie));
+
+	return ret;
+}
+
+/**
+ * Pseudo random source for Mbed
+ *
+ * @param data    Optional entropy source.
+ * @param output  Buffer to receive random numbers.
+ * @param len     Number of bytes requested.
+ * @param olen    Number of bytes actually returned.
+ *
+ * @return Always zero.
+ */
+static int auth_tls_entropy(void *data, unsigned char *output, size_t len,
+			    size_t *olen)
+{
+	(void) data;
+
+	sys_rand_get(output, len);
+	*olen = len;
+
+	return 0;
+}
+
+
+
+/* ================= external/internal funcs ==================== */
+/**
+ * @see auth_internal.h
+ *
+ */
+int auth_init_dtls_method(struct authenticate_conn *auth_conn, struct auth_dtls_certs *certs)
+{
+	struct mbed_tls_context *mbed_ctx;
+	int ret;
+
+	LOG_DBG("Initializing Mbed DTLS");
+
+	/* get, init, and set context pointer */
+	mbed_ctx = auth_get_mbedcontext();
+
+	if (mbed_ctx == NULL) {
+		LOG_ERR("Unable to allocate Mbed TLS context.");
+		return AUTH_ERROR_NO_RESOURCE;
+	}
+
+	/* Init mbed context */
+	auth_init_context(mbed_ctx);
+
+
+	/* Save MBED tls context as internal object. The intent of using a void
+	 * 'internal_obj' is to provide a var in the struct authentication_conn to
+	 * store different authentication methods.  Instead of Mbed, it maybe a
+	 * Challenge-Response.*/
+	auth_conn->internal_obj = mbed_ctx;
+
+	int endpoint = auth_conn->is_client ? MBEDTLS_SSL_IS_CLIENT : MBEDTLS_SSL_IS_SERVER;
+
+	mbedtls_ssl_config_defaults(&mbed_ctx->conf,
+				    endpoint,
+				    MBEDTLS_SSL_TRANSPORT_DATAGRAM,
+				    MBEDTLS_SSL_PRESET_DEFAULT);
+
+
+	/* set the lower layer transport functions */
+	mbedtls_ssl_set_bio(&mbed_ctx->ssl, auth_conn, auth_mbedtls_tx, auth_mbedtls_rx, NULL);
+
+	/* set max record len to 512, as small as possible */
+	mbedtls_ssl_conf_max_frag_len(&mbed_ctx->conf, MBEDTLS_SSL_MAX_FRAG_LEN_512);
+
+	/* Set the DTLS time out */
+	mbedtls_ssl_conf_handshake_timeout(&mbed_ctx->conf, AUTH_DTLS_MIN_TIMEOUT,
+					   ATUH_DTLS_MAX_TIMEOUT);
+
+	/*  Force verification.  */
+	mbedtls_ssl_conf_authmode(&mbed_ctx->conf, MBEDTLS_SSL_VERIFY_REQUIRED);
+
+
+	if ((certs->device_cert_pem.priv_key == NULL) || (certs->device_cert_pem.priv_key_size == 0)) {
+		auth_free_mbedcontext(mbed_ctx);
+		LOG_ERR("Failed to get device private key");
+		return AUTH_ERROR_DTLS_INIT_FAILED;
```

Comment:
this pattern is repetitive across this function, just log the problem and goto an error label

Reply:
Good point.  Added goto error label.